Add task trace support

YuuichiNakamura opened the pull request:
## Summary
This PR adds "trace" command for task trace.

- Currently the documentation is available at https://github.com/YuuichiNakamura/nuttx-task-tracer-doc/blob/master/NuttXTaskTracer.md 
- rst version will be prepared later.

## Impact
- "trace" command is added as built-in apps. It is available only when the related kernel configurations are enabled.

## Testing
Tested by spresense:nsh by changing the following configurations.
```
CONFIG_SCHED_INSTRUMENTATION=y
CONFIG_SCHED_INSTRUMENTATION_FILTER=y
CONFIG_SCHED_INSTRUMENTATION_SYSCALL=y
CONFIG_SCHED_INSTRUMENTATION_IRQHANDLER=y
CONFIG_DRIVER_NOTE=y
CONFIG_DRIVER_NOTERAM=y
CONFIG_DRIVER_NOTECTL=y
CONFIG_SYSTEM_TRACE=y
CONFIG_SYSTEM_SYSTEM=y
```
Then "trace" command is added as built-in apps.
```
nsh> trace cmd usleep 1000
nsh> trace dump
```
This can get the following results:
```
<noname>-1   [0]   7.640000000: sys_ioctl -> 0
<noname>-1   [0]   7.640000000: sys_close()
<noname>-1   [0]   7.640000000: sys_close -> 0
<noname>-1   [0]   7.640000000: sys_sched_lock()
<noname>-1   [0]   7.640000000: sys_sched_lock -> 0
<noname>-1   [0]   7.640000000: sys_nxsched_get_stackinfo()
<noname>-1   [0]   7.640000000: sys_nxsched_get_stackinfo -> 0
<noname>-1   [0]   7.640000000: sys_sched_unlock()
<noname>-1   [0]   7.640000000: sys_sched_unlock -> 0
<noname>-1   [0]   7.640000000: sys_clock_nanosleep()
<noname>-1   [0]   7.640000000: sched_switch: prev_comm=<noname> prev_pid=1 prev_state=S ==> next_comm=<noname> next_pid=0
<noname>-0   [0]   7.640000000: irq_handler_entry: irq=11
<noname>-0   [0]   7.640000000: irq_handler_exit: irq=11
<noname>-0   [0]   7.640000000: irq_handler_entry: irq=15
<noname>-0   [0]   7.650000000: irq_handler_exit: irq=15
<noname>-0   [0]   7.650000000: irq_handler_entry: irq=15
<noname>-0   [0]   7.660000000: sched_waking: comm=<noname> pid=1 target_cpu=0
<noname>-1   [0]   7.660000000: irq_handler_exit: irq=15
<noname>-1   [0]   7.660000000: sched_switch: prev_comm=<noname> prev_pid=1 prev_state=R ==> next_comm=<noname> next_pid=1
<noname>-1   [0]   7.660000000: sys_clock_nanosleep -> 0
```

patacongo commented:
I don't believe that this should be a part of NSH.  NSH is a thin shell, like Bash, with a few simple built-in functions but its primary purpose is to interact with the system and with other programs.  It should NOT contain any specialized programs itself.  Any specialized programs must be moved to apps/system as built-in applications.  There is no difference from the user experience, but this properly prevents NSH from exploding in size with specialized, programs.  It must stay lean.

YuuichiNakamura commented:
> I don't believe that this should be a part of NSH. NSH is a thin shell, like Bash, with a few simple built-in functions but its primary purpose is to interact with the system and with other programs. It should NOT contain any specialized programs itself. Any specialized programs must be moved to apps/system as built-in applications. There is no difference from the user experience, but this properly prevents NSH from exploding in size with specialized, programs. It must stay lean.

@patacongo I understood. I'll try to fix it to built-in apps.

YuuichiNakamura commented:
@patacongo I have fixed the implementation as the built-in apps and moved into apps/system/trace.
@xiaoxiang781216 Some of your comments were solved by changing into built-in apps.

YuuichiNakamura commented on nshlib/nsh_command.c:
```diff
@@ -518,6 +518,18 @@ static const struct cmdmap_s g_cmdmap[] =
 
+#ifdef CONFIG_DRIVER_NOTECTL
+# ifndef CONFIG_NSH_DISABLE_TRACE
+  { "trace",    cmd_trace,    1, CONFIG_NSH_MAXARGUMENTS,
```
To say the truth, I troubled what command name should be given to this feature.
As you mentioned, this feature strongly depends on note of the kernel function, but "note" command already exists at apps/system/sched_note, and from the user point of view, the feature realized by this command is truly task trace.
So I decided to name "trace" to this feature.

xiaoxiang781216 replied:
Sure, but your tool is a superset of sched_note, should we merge both into one to reduce the maintainence cost? BTW, sched_node may already in broken state with the recent note change.

YuuichiNakamura replied:
sched_note shows the scheduler notes as it is and I think it is important for kernel development, especially for scheduler.
Different to it, my command processes and replaces the notes to fit for tracecompass. Moreover, basically my command intends to handle the notes after the trace is stopped. This is a different behavior to sched_note. So I'm negative to merge them into one.
sched_note is still working. My changes for note device doesn't affect the existing behavior.

xiaoxiang781216 replied:
Ok.

YuuichiNakamura commented:
The last commit 7e3a54a is a very trivial changes. Only README.md was changed.

xiaoxiang781216 commented on system/trace/trace_dump.c:
```diff
@@ -0,0 +1,570 @@
+ ****************************************************************************/
+
+static struct trace_dump_cpu_context_s g_dump_cpu_ctx[NCPUS];
+static FAR struct trace_dump_task_context_s *g_dump_task_ctx = NULL;
```
let's change to local variable?

YuuichiNakamura replied:
You mean both g_dump_cpu_ctx[] and g_dump_task_ctx?
It seems a bit complicated to add both argument to every private functions.
I think we have no strong reason to change into local variable because:
- These are widely accessed by private functions in this file, 
- Already defined as static and used only in this file.
- The function never called recursively.

xiaoxiang781216 commented on system/trace/trace_dump.c:
```diff
@@ -0,0 +1,570 @@
+          get_task_name(pid), get_pid(pid), cpu,
+          systime / (1000 * 1000 / CONFIG_USEC_PER_TICK),
+          (systime % (1000 * 1000 / CONFIG_USEC_PER_TICK))
+            * CONFIG_USEC_PER_TICK * 1000);
```
why need multiple by CONFIG_USEC_PER_TICK?

YuuichiNakamura replied:
`systime % (1000 * 1000 / CONFIG_USEC_PER_TICK)` gets the number of ticks less than a second.
Because it is a unit of ticks, we need to multiply `CONFIG_USEC_PER_TICK * 1000` to get a unit of nanosecs which the ftrace format requires.

xiaoxiang781216 commented on system/trace/trace.c:
```diff
@@ -0,0 +1,812 @@
+  int exitcode;
+  int i;
+
+  exitcode = EXIT_FAILURE;
```
merge to line 727?

xiaoxiang781216 commented on system/trace/trace.c:
```diff
@@ -0,0 +1,812 @@
+ ****************************************************************************/
+
+#ifdef CONFIG_SCHED_INSTRUMENTATION_IRQHANDLER
+static int trace_cmd_irq(int index, int argc, char **argv)
```
add FAR to all pointer type?

xiaoxiang781216 commented on system/trace/trace.c:
```diff
@@ -0,0 +1,812 @@
+                                " Get the trace while running <command>\n"
+#endif
+#ifdef CONFIG_DRIVER_NOTERAM
+          "  dump [<filename>]               :"
```
Since trace without argument is same as trace dump wihtout filename, how about we move {+|-}o to here to centralize all dump option in one place? e.g.:
dump ->show the overwrite state
dump filename->save to file
dump +o->overwrite
dump -o->don't ovwerwrite
The benefit is that, we can add function to trace_dump.c like:
trace_dump_cmd
trace_dump_show_usage
and trace.c just forward to these function.

YuuichiNakamura replied:
I tried to change trace_cmd_dump() to move overwrite option from trace_cmd_mode().
But, I think it should not be moved into trace_dump.c. Because,
- The command line parsing is split into two files.
- trace_cmd_dump() needs to call notectl_enable() because https://github.com/apache/incubator-nuttx-apps/pull/421#discussion_r502805629 . Then, both files need to call another mutually and it becomes not simple.

YuuichiNakamura replied:
Though once I have tried to fix "trace dump" command to follow your proposals, I have reverted the changes.
Especially, the following syntax seems strange to me.
trace -> dump trace data to console
trace dump ->show the overwrite state
trace dump filename->save to file

I have reverted it and changed the default behavior of trace command (without options).
trace -> show current state (same as "trace mode")
trace dump -> dump trace data to console
trace dump filename -> save to file (dump trace data to file)
How about it?

xiaoxiang781216 replied:
No problem.

xiaoxiang781216 commented on system/trace/trace.c:
```diff
@@ -0,0 +1,812 @@
+
+  notectl_enable(true);
+  system(command);
+  notectl_enable(false);
```
should we restore the old state?

YuuichiNakamura commented on system/trace/trace.h:
```diff
@@ -0,0 +1,59 @@
+ *
+ ****************************************************************************/
+
+int trace_dump(FILE *out);
```
If the whole file is guarded in Makefile (https://github.com/apache/incubator-nuttx-apps/pull/421#discussion_r502535529), I think we have no need to guard here.

YuuichiNakamura commented on system/trace/trace.c:
```diff
@@ -0,0 +1,812 @@
+  struct note_filter_mode_s mode;
+
+#ifdef CONFIG_BUILD_FLAT
+  sched_note_filter_mode(&mode, NULL);
```
It is to avoid to record the ioctl which controls a trace mode.

YuuichiNakamura replied:
That's true... I didn't like remaining an irrelevant ioctls in log but consistency is important.

YuuichiNakamura commented on system/trace/trace.c:
```diff
@@ -0,0 +1,812 @@
+ ****************************************************************************/
+
+#ifdef CONFIG_DRIVER_NOTERAM
+static void note_ioctl(int cmd, unsigned long arg)
```
I got it. But line 35-37 cannot be removed because NOTERAM_MODE_OVERWRITE defines are still required.

xiaoxiang781216 replied:
How about we change trace_dump_get_mode to trace_dump_get_overwrite or trace_dump_is_overwrite? so true/false is used in trace.c and trace_dump.c convert true/false to NOTERAM_MODE_OVERWRITE?

YuuichiNakamura commented on system/trace/trace.c:
```diff
@@ -0,0 +1,812 @@
+
+  /* Stop the tracing before dump */
+
+  notectl_enable(false);
```
trace_dump() continues dumping until all data in the buffer is dumped. 
Because trace_dump() itself issues syscalls, if the trace is not stopped before dump, these are added in the buffer while dumping. So, once the dump is started, it never stops.

YuuichiNakamura commented on system/trace/trace.c:
```diff
@@ -0,0 +1,812 @@
+      /* Enable irq trace flag */
+
+      ioctl(g_notectlfd, NOTECTL_GETMODE, (unsigned long)&mode);
+      mode.flag |= NOTE_FILTER_MODE_FLAG_IRQ;
```
I got it. I have changed to let user invoke mode command manually.

YuuichiNakamura commented on system/trace/trace.c:
```diff
@@ -0,0 +1,812 @@
+    {
+      /* No arguments - dump the trace data */
+
+      notectl_enable(false);
```
Same as https://github.com/apache/incubator-nuttx-apps/pull/421#discussion_r503734494

YuuichiNakamura commented on system/trace/trace.c:
```diff
@@ -0,0 +1,812 @@
+      /* Enable syscall trace flag */
+
+      ioctl(g_notectlfd, NOTECTL_GETMODE, (unsigned long)&mode);
+      mode.flag |= NOTE_FILTER_MODE_FLAG_SYSCALL;
```
Same as https://github.com/apache/incubator-nuttx-apps/pull/421#discussion_r503734715

YuuichiNakamura commented on system/trace/trace.c:
```diff
@@ -0,0 +1,751 @@
+#include <nuttx/lib/regex.h>
+#include <nuttx/note/notectl_driver.h>
+#ifdef CONFIG_DRIVER_NOTERAM
+#  include <nuttx/note/noteram_driver.h>
```
It is still needed because NOTERAM_MODE_OVERWRITE_ENABLE/DISABLE definition is used in trace_cmd_mode().

xiaoxiang781216 replied:
How about we change trace_dump_get_mode to trace_dump_get_overwrite or trace_dump_is_overwrite? so true/false is used in trace.c and trace_dump.c convert true/false to NOTERAM_MODE_OVERWRITE?

YuuichiNakamura commented on system/trace/trace.c:
```diff
@@ -0,0 +1,751 @@
+  /* Clear the trace buffer */
+
+#ifdef CONFIG_DRIVER_NOTERAM
+  trace_dump_clear();
```
Adding "-c" (continue trace) option seems good.

YuuichiNakamura commented on system/trace/trace.c:
```diff
@@ -0,0 +1,751 @@
+      return ERROR;
+    }
+
+  /* Stop the tracing before dump */
```
I'd like to add a similar option like "trace start -c".

YuuichiNakamura commented on system/trace/trace_dump.c:
```diff
@@ -0,0 +1,643 @@
+#endif
+     )
+    {
+      cctx->current_pid = pid;
```
It is needed because cctx->current_pid isn't updated until the first context switch in the trace.
The trace data begins in incomplete state and the beginning of the trace dump also shows incomplete result. This line fixes such situation though it is not perfect.

YuuichiNakamura commented on system/trace/trace_dump.c:
```diff
@@ -0,0 +1,643 @@
+              break;
+            }
+
+          tctx->syscall_nest = 0;
```
It is needed because syscall enter and leave in the trace data are not always coupled.
If only syscall enter is overwritten and lost in the trace data, the first syscall leave may make nest to negative value.

YuuichiNakamura commented on system/trace/trace_dump.c:
```diff
@@ -0,0 +1,643 @@
+            }
+
+          tctx->syscall_nest++;
+          if (tctx->syscall_nest > 1)
```
The syscall nest while executing the interrupt handler should not belong to the syscall nest in the task.
For now, because all syscalls in the interrupt handler cannot be displayed, we can ignore the syscall immediately if in the interrupt handler.

xiaoxiang781216 commented on system/trace/trace.c:
```diff
@@ -0,0 +1,775 @@
+
+  /* Clear the trace buffer */
+
+#ifdef CONFIG_DRIVER_NOTERAM
```
how about we change trace.h like this:
```
#ifdef CONFIG_DRIVER_NOTERAM
void trace_dump_clear(void);
#else
#define trace_dump_clear
#endif
```
so we don't need check CONFIG_DRIVER_NOTERAM again and again. other functions in trace.h could apply the approach.

xiaoxiang781216 commented on system/trace/trace.c:
```diff
@@ -0,0 +1,775 @@
+
+  ret = trace_dump(out);
+
+  if (!cont && changed)
```
let's change line 152 to: bool changed = false and remove !cont &&?

xiaoxiang781216 commented on system/trace/trace.c:
```diff
@@ -0,0 +1,775 @@
+#ifdef CONFIG_DRIVER_NOTERAM
+  if (!cont)
+    {
+      trace_dump_clear();
```
same comment as line 120

xiaoxiang781216 commented on system/trace/trace.c:
```diff
@@ -0,0 +1,775 @@
+
+  ioctl(notectlfd, NOTECTL_GETMODE, (unsigned long)&mode);
+#ifdef CONFIG_DRIVER_NOTERAM
+  owmode = trace_dump_getmode();
```
same comment as line 120

xiaoxiang781216 reviewed (approved):
LGTM.

YuuichiNakamura commented:
@xiaoxiang781216 Thank you for your great support for my kernel and app PRs!
I'll work on the maintenance of the documents in https://github.com/apache/incubator-nuttx/pull/1955 .
